protostar/compiler/contract_source_identifier.py

```diff
@@ -0,0 +1,65 @@
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Protocol
+
+from protostar.protostar_exception import ProtostarException
+from protostar.self import ContractName
+
+
+@dataclass(eq=True, frozen=True, unsafe_hash=True)
+class ContractSourceIdentifier:
+    @classmethod
+    def from_contract_path(cls, path: Path):
+        return cls(
+            name=path.stem,
+            paths=[path],
+        )
+
+    @classmethod
+    def create(cls, name: ContractName, paths: list[Path]):
```

Comment:
Why we reimplement constructor under a different name? It is confusing. Cannot we just validate dataclass post_init?

Reply:
Constructors shouldn't raise errors. Static factories can.

Reply:
What is the reason behind this rule 🤔

Reply:
Constructors shouldn't throw errors because they shouldn't include business logic. They shouldn't include business logic to make the composition fast and delay potential errors. In case of data structures, this is done to preserve consistency.

Moreover, functions/methods should do one thing and only one thing. I/O operations shouldn't be constructor responsibility.